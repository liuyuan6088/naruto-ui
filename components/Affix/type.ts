
export interface IAffixProps {
  // 距离窗口顶部达到指定偏移量后触发
  offsetTop?: number
  // 距离窗口底部达到指定偏移量后触发
  offsetBottom?: number
  // 容器
  target?: () => HTMLElement
  // 固定状态改变时触发的回调函数
  onChange?: (affixed: boolean) => void
  // 类名
  className?: string
  // 样式
  style?: React.CSSProperties
}