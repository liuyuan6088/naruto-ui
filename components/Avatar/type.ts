
import { ICommonProps, tuple } from '../utils/type';

export const AvatarSizes = tuple('large', 'default', 'small');
export type AvatarSize = (typeof AvatarSizes)[number];

export const AvatarShape = tuple('circle', 'square');
export type AvatarShape = (typeof AvatarShape)[number];

export interface IAvatarProps extends ICommonProps {
  children?: React.ReactNode;
  // 设置头像的图标类型，参考 Icon 组件
  icon?: string;
  // 设置头像的大小
  size?: number | AvatarSize;
  // 指定头像的形状
  shape?: AvatarShape;
  // 图片类头像的资源地址
  src?: string;
  // 设置图片类头像响应式资源地址
  srcSet?: string;
  // 图像无法显示时的替代文本
  alt?: string;
  // 图片加载失败的事件，返回 false 会关闭组件默认的 fallback 行为
  onError?: () => boolean;
  // 点击事件
  onClick?: (e: React.MouseEvent) => void;
}