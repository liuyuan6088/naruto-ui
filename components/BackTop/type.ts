
import { ICommonProps } from '../utils/type';

export interface IBackTopProps extends ICommonProps {
  // 滚动高度达到此参数值才出现 BackTop  默认400
  visibilityHeight?: number;
  // 设置需要监听其滚动事件的元素，值为一个返回对应 DOM 元素的函数
  target?: () => (HTMLElement | Window);
  // 点击按钮的回调函数
  onClick?: (e: React.MouseEvent) => void;
}