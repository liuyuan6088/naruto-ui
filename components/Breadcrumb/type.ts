
import { ICommonProps } from '../utils/type';

export interface Route {
  path: string;
  breadcrumbName: string;
}

export interface IBreadcrumbItemProps {
  // 分隔符自定义
  separator?: string | React.ReactNode;
  // 类名前缀
  prefixCls?: string;
  // 链接的目的地
  href?: string;
  // 单击事件
  onClick?: (e: any)=>void;
  // 是否最后一个
  isLast?: boolean;
}

export interface IBreadcrumbProps extends ICommonProps {
  // router 的路由栈信息
  routes?: Route[];
  //  路由的参数
  params?: object;
  // 分隔符自定义
  separator?: string | React.ReactNode;
  // 自定义链接函数
  itemRender?: (
    route: any,
    params: any,
    routes: Array<any>,
    paths: Array<string>,
  ) => React.ReactNode;
}