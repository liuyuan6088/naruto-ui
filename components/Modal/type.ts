
import { IButtonProps } from '../Button/type';
import { ICommonProps, Omit } from '../utils/type';
// import { CSSTimer } from '../Overlay/type';

export interface IModalProps extends ICommonProps {

  visible?: boolean;
  title?: string | React.ReactNode;

  // 取消文本 默认为 `取消`
  cancelText?: string | React.ReactNode;
  // okText 确认文本 默认为 `确定`
  okText?: string | React.ReactNode;
  // onOk 点击确定的回调
  onOk?: () => void;
  // onCancel 点击取消或遮罩层或者右上角按钮的回调
  onCancel?: () => void;
  // okButtonProps
  okButtonProps?: IButtonProps;
  // 取消按钮的 props
  cancelButtonProps?: IButtonProps;
  // footer 底部 当不需要时设置 null
  footer?: React.ReactNode | null;

  // closable 是否显示右上角的叉号 默认 `true`
  closable?: boolean;
  // destroy 隐藏时， 是否销毁 默认为 true
  destroy?: boolean;
  // esc 是否支持键盘 esc 关闭 默认为 true
  esc?: boolean;
  // maskClosable 点击蒙层是否关闭弹窗 默认为 `true`
  maskClosable?: boolean;
  // zIndex 设置 modal 的 z-index
  zIndex?: number;
  // width 可主动设置内容区的宽度
  width?: number;
  // afterClose 弹窗完全关闭的回调
  afterClose?: () => void;
  // center 是否居中显示 modal 默认为 false
  center?: boolean;
  // 确定按钮 loading
  confirmLoading?: boolean;
  // 居中
  centered?: boolean;
}

export interface IAlertProps
  extends Omit<IModalProps, 'visible' | 'footer' | 'closable' | 'destroy'> {
  /** icon 类型 可自定义 默认与 快捷弹窗的名称一致 */
  icon?: React.ReactNode
  /** content 内容 */
  content?: string | React.ReactNode
  /** hasCancelBtn 是否需要取消按钮 默认为 true */
  hasCancelBtn?: boolean
}