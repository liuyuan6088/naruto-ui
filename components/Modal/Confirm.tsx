import * as React from 'react'
import * as ReactDOM from 'react-dom'
import cx from 'classnames'
import Modal from './Modal'
import Button from '../Button'
import Icon from '../Icon'
import { primaryName } from '../utils/constant'
import { IAlertProps } from './type'
import './style/index.less'

const noop = () => {}
const prefixCls = `${primaryName}-modal-confirm`
export interface Close {
  close: () => void
}

const Confirm = ({
  content,
  icon = <Icon type="question-circle" color="#faad14" />,
  okText = '确认',
  onOk = noop,
  okButtonProps,
  title,
  onCancel = noop,
  cancelButtonProps,
  cancelText = '取消',
  className,
  hasCancelBtn = true,
  ...rest
}: IAlertProps): Close => {
  let isClosed = false

  const div = document.createElement('div')
  document.body.appendChild(div)

  const classes = cx(`${prefixCls}`, className)

  const close = () => {
    ReactDOM.render(renderModal(false), div)
  }

  const destroy = () => {
    ReactDOM.unmountComponentAtNode(div)
    if (div && div.parentNode) {
      div.parentNode.removeChild(div)
    }
  }

  const handleClick = () => {
    if (isClosed) return
    Promise.resolve(onOk())
      .then(() => {
        isClosed = true
        close()
      })
      .catch(() => {
        isClosed = true
        close()
      })
  }

  const renderModal = (visible: boolean) => {
    return (
      <Modal
        {...rest}
        closable={false}
        onCancel={() => {
          onCancel()
          close()
        }}
        title={null}
        footer={null}
        visible={visible}
        afterClose={destroy}
        className={classes}
      >
        <div className={`${prefixCls}-wrapper`}>
          <div className={`${prefixCls}-body`}>
            <div className={`${prefixCls}-body-header`}>
              {icon}
              {title ? <div className={`${prefixCls}-title`}>{title}</div> : null}
            </div>
            {content ? <div className={`${prefixCls}-content`}>{content}</div> : null}
          </div>
          <div className={`${prefixCls}-btn`}>
            {hasCancelBtn ? (
              <Button
                onClick={() => {
                  onCancel()
                  close()
                }}
                {...cancelButtonProps}
              >
                {cancelText}
              </Button>
            ) : null}
            <Button type="primary" onClick={handleClick} {...okButtonProps}>
              {okText}
            </Button>
          </div>
        </div>
      </Modal>
    )
  }

  ReactDOM.render(renderModal(true), div)

  return {
    close
  }
}

export default Confirm

export function info({
  icon = <Icon type="info-circle" color="#1890ff" />,
  hasCancelBtn = false,
  okText = '知道了',
  ...rest
}: IAlertProps) {
  return Confirm({ icon, hasCancelBtn, okText, ...rest })
}

export function success({
  icon = <Icon type="check-circle" color="#52c41a" />,
  hasCancelBtn = false,
  okText = '知道了',
  ...rest
}: IAlertProps) {
  return Confirm({ icon, hasCancelBtn, okText, ...rest })
}

export function error({
  icon = <Icon type="close-circle" color="#f5222d" />,
  hasCancelBtn = false,
  okText = '知道了',
  ...rest
}: IAlertProps) {
  return Confirm({ icon, hasCancelBtn, okText, ...rest })
}

export function warning({
  icon = <Icon type="warning-circle" color="#faad14" />,
  hasCancelBtn = false,
  okText = '知道了',
  ...rest
}: IAlertProps) {
  return Confirm({ icon, hasCancelBtn, okText, ...rest })
}
