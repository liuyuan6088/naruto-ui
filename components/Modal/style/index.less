@import '../../styles/theme/default.less';

@modal-prefix-cls: ~"@{primary-name}-modal";

.@{modal-prefix-cls} {

  position: fixed;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  z-index: @modalIndex;
  outline: 0;

  &-center {
    text-align: center;
  }

  &-wrapper {
    background: #fff;
    border-radius: @bordor-radius-1;
    min-width: 500px;
    color: rgba(0, 0, 0, 0.65);
    position: absolute;
    left: 50%;
    transform: translateX(-50%);

    &-close {
      position: absolute;
      top: @space-0;
      right: @space-1;
      cursor: pointer;
      transition: all .2s ease;
      // &:hover {
      //   transform: rotate(90deg)
      // }
    }

    &-header {
      padding: @space-0 @space-1;
      border-bottom: 1px solid @border-color-0;
      .@{modal-prefix-cls}-title {
        font-size: @font-size-large;
        font-weight: 700;
        color: @gray-2;
      }
    }

    &-content {
      padding: @space-1;
      font-size: @font-size-small;
      line-height: 1;
      word-wrap: break-word;
    }

    &-footer {
      padding: @space-sm @space-0;
      border-top: 1px solid @border-color-0;
      display: flex;
      justify-content: flex-end;
      button {
        + button {
          margin-left: 15px;
        }
      }
    }
  }

  &-mask {
    // width: 100vw;
    // height: 100vh;
    position: fixed;
    left: 0;
    right: 0;
    top: 0;
    bottom: 0;
    background: @maskBgc;
    z-index: @modalIndex;
  }

  &-confirm {
    &-wrapper {
      min-width: 400px;
    }
  
    &-wrapper-inner {
      padding: 32px 32px 24px;
    }
  
    &-body-header {
      display: flex;
      align-items: center;
      padding: @space-sm;
      > svg {
        font-size: 24px;
      }
    }
  
    &-title {
      margin-left: 10px;
      color: @gray-2;
      font-weight: 700;
    }
  
    &-content {
      padding: 0 45px;
      margin: 15px 0;
      color: @gray-0;
    }
  
    &-btn {
      text-align: right;
      > button {
        + button {
          margin-left: 20px;
        }
      }
    }
  }

  //
  &-fade-appear,
  &-fade-enter {
    opacity: 0;
    animation-duration: 0.3s;
    animation-fill-mode: both;
    animation-timing-function: linear;
    animation-play-state: paused;
  }

  &-fade-appear-active,
  &-fade-enter-active {
    animation-name: modalFadeIn;
    animation-play-state: running;
  }

  &-fade-exit {
    animation-duration: 0.3s;
    animation-fill-mode: both;
    animation-timing-function: linear;
    animation-play-state: paused;
  }

  &-fade-exit-active {
    animation-name: modalFadeOut;
    animation-play-state: running;
  }

  &-zoom-appear,
  &-zoom-enter {
    opacity: 0;
    animation-duration: 0.3s;
    animation-fill-mode: both;
    animation-timing-function: @ease-out-circ;
    animation-play-state: paused;
  }

  &-zoom-appear-active,
  &-zoom-enter-active {
    animation-name: modalZoomIn;
    animation-play-state: running;
  }

  &-zoom-exit {
    animation-duration: 0.3s;
    animation-fill-mode: both;
    animation-timing-function: @ease-out-circ;
    animation-play-state: paused;
  }

  &-zoom-exit-active {
    animation-name: modalZoomOut;
    animation-play-state: running;
  }

}

@keyframes modalFadeIn {
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
}

@keyframes modalFadeOut {
  0% {
    opacity: 1;
  }

  100% {
    opacity: 0;
  }
}

@keyframes modalZoomIn {
  0% {
    opacity: 0;
    transform: scale(0.2);
  }

  100% {
    opacity: 1;
    transform: scale(1);
  }
}

@keyframes modalZoomOut {
  0% {
    transform: scale(1);
  }

  100% {
    opacity: 0;
    transform: scale(0.2);
  }
}