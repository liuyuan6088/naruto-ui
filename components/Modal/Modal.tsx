import * as React from 'react'
import cx from 'classnames'
import Overlay from '../Overlay'
import Button from '../Button'
import Confirm, { info, success, error, warning } from './Confirm'
import { IModalProps } from './type'
import { primaryName } from '../utils/constant'
import './style/index.less'

const noop = () => {}

const defaultProps: IModalProps = {
  prefixCls: `${primaryName}-modal`,
  cancelText: '取消',
  okText: '确定',
  onOk: noop,
  onCancel: noop,
  visible: false,
  closable: true,
  destroy: true,
  esc: true,
  style: {},
  confirmLoading: false,
  maskClosable: true
}

const renderHeader = ({ title, prefixCls }: IModalProps) => {
  if (title) {
    return <div className={`${prefixCls}-title`}>{title}</div>
  }
  return null
}

const renderFooter = ({
  cancelText,
  okText,
  onOk,
  okButtonProps,
  cancelButtonProps,
  footer,
  onCancel,
  confirmLoading
}: IModalProps) => {
  if (footer === null) return null
  return (
    <div>
      {footer || (
        <React.Fragment>
          <Button onClick={onCancel} {...cancelButtonProps}>
            {cancelText}
          </Button>
          <Button loading={confirmLoading} type="primary" onClick={onOk} {...okButtonProps}>
            {okText}
          </Button>
        </React.Fragment>
      )}
    </div>
  )
}

const Modal: React.FC<IModalProps> &
  Record<'confirm' | 'success' | 'error' | 'warning' | 'info', typeof Confirm> = props => {
  const {
    visible,
    children,
    prefixCls,
    closable,
    onCancel,
    destroy,
    esc,
    maskClosable,
    zIndex,
    className,
    style,
    width,
    center,
    afterClose
  } = props

  const classes = cx(
    {
      [`${prefixCls}-center`]: center
    },
    className
  )

  const widthStyle = width ? { width } : {}

  return (
    <Overlay
      visible={visible}
      prefixCls={prefixCls}
      wrapperStyle={{ ...widthStyle, ...style }}
      wrapperClassName={classes}
      footer={renderFooter(props)}
      header={renderHeader(props)}
      closable={closable}
      onClose={onCancel}
      destroy={destroy}
      esc={esc}
      maskClosable={maskClosable}
      zIndex={zIndex}
      afterClose={afterClose}
    >
      {children}
    </Overlay>
  )
}

Modal.defaultProps = defaultProps
Modal.confirm = Confirm
Modal.success = success
Modal.error = error
Modal.warning = warning
Modal.info = info

export default Modal
