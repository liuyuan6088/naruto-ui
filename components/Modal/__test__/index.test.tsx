import * as React from 'react'
import { mount } from 'enzyme'
import Modal from '../index'
import { IModalProps } from '../type'

const { useState, useEffect } = React

const Demo: React.FC<IModalProps> = props => {
  const [visible1, setVisible1] = useState<boolean>(false)

  useEffect(() => {
    setVisible1(true)
  }, [])

  return (
    <React.Fragment>
      <p>最简单的用法</p>
      <div className="demo-modal">
        <Modal visible={visible1} title="modal title" {...props}>
          modal content
        </Modal>
      </div>
    </React.Fragment>
  )
}

describe('render modal', () => {
  it('render correctly', () => {
    const wrapper = mount(<Demo />)
    expect(wrapper.render()).toMatchSnapshot()
  })

  it('render without footer', () => {
    const wrapper = mount(<Demo footer={null} />)
    expect(wrapper.render()).toMatchSnapshot()
  })
})
