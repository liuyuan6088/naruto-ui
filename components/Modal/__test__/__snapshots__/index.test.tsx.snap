// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render modal render correctly 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "data": "最简单的用法",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "p",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": Object {
      "attribs": Object {
        "class": "demo-modal",
      },
      "children": Array [],
      "name": "div",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": [Circular],
      "root": Object {
        "attribs": Object {},
        "children": Array [
          [Circular],
          [Circular],
        ],
        "name": "root",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": null,
        "prev": null,
        "type": "root",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
      "type": "tag",
      "x-attribsNamespace": Object {
        "class": undefined,
      },
      "x-attribsPrefix": Object {
        "class": undefined,
      },
    },
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
        Object {
          "attribs": Object {
            "class": "demo-modal",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": null,
          "prev": [Circular],
          "root": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "1": Object {
    "attribs": Object {
      "class": "demo-modal",
    },
    "children": Array [],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": Object {
      "attribs": Object {},
      "children": Array [
        Object {
          "data": "最简单的用法",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "text",
        },
      ],
      "name": "p",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": [Circular],
      "parent": null,
      "prev": null,
      "root": Object {
        "attribs": Object {},
        "children": Array [
          [Circular],
          [Circular],
        ],
        "name": "root",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": null,
        "prev": null,
        "type": "root",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
      "type": "tag",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "root": Object {
      "attribs": Object {},
      "children": Array [
        Object {
          "attribs": Object {},
          "children": Array [
            Object {
              "data": "最简单的用法",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "p",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": null,
          "prev": null,
          "root": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 2,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`render modal render without footer 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "data": "最简单的用法",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "p",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": Object {
      "attribs": Object {
        "class": "demo-modal",
      },
      "children": Array [],
      "name": "div",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": [Circular],
      "root": Object {
        "attribs": Object {},
        "children": Array [
          [Circular],
          [Circular],
        ],
        "name": "root",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": null,
        "prev": null,
        "type": "root",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
      "type": "tag",
      "x-attribsNamespace": Object {
        "class": undefined,
      },
      "x-attribsPrefix": Object {
        "class": undefined,
      },
    },
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
        Object {
          "attribs": Object {
            "class": "demo-modal",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": null,
          "prev": [Circular],
          "root": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "1": Object {
    "attribs": Object {
      "class": "demo-modal",
    },
    "children": Array [],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": Object {
      "attribs": Object {},
      "children": Array [
        Object {
          "data": "最简单的用法",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "text",
        },
      ],
      "name": "p",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": [Circular],
      "parent": null,
      "prev": null,
      "root": Object {
        "attribs": Object {},
        "children": Array [
          [Circular],
          [Circular],
        ],
        "name": "root",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": null,
        "prev": null,
        "type": "root",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
      "type": "tag",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "root": Object {
      "attribs": Object {},
      "children": Array [
        Object {
          "attribs": Object {},
          "children": Array [
            Object {
              "data": "最简单的用法",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "p",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": null,
          "prev": null,
          "root": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 2,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
